File: Top200/144-PreorderTraversal.java
package Top200;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

class PreorderTraversal {
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) {
            val = x;
        }
    }    

    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        if (root == null) {
            return res;
        }

        getPre(root, res);
        return res;
    }

    public void getPre(TreeNode node, List<Integer> res) {
        res.add(node.val);

        if (node.left != null) {
            getPre(node.left, res);
        }

        if (node.right != null) {
            getPre(node.right, res);
        }
    }

    public List<Integer> preorderTraversal2(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        Stack<TreeNode> stack = new Stack<TreeNode>();

        if (root == null) {
            return res;
        }

        stack.push(root);

        while (!stack.isEmpty()) {
            TreeNode node = stack.pop();
            res.add(node.val);

            if (node.right != null) {
                stack.push(node.right);
            }

            if (node.left != null) {
                stack.push(node.left);
            }
        }
        
        return res;
    }

    public List<Integer> preorderTraversal3(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        
        TreeNode curr = root;
        TreeNode pre = null;

        while (curr != null) {
            if (curr.left == null) {
                res.add(curr.val);
                curr = curr.right;     
            } else {
                pre = curr.left;
                while (pre.right != null && pre.right != curr) {
                    pre = pre.right;
                }

                if (pre.right ==  null) {
                    res.add(curr.val);
                    pre.right = curr;
                    curr = curr.left;
                } else {
                    pre.right = null;
                    curr = curr.right;
                }
            }
        }

        return res;
    }
}

File: Top200/144-PreorderTraversal.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type Stack struct {
	arr  []*TreeNode
	size int
}

func NewStack() *Stack {
	return &Stack{
		arr:  []*TreeNode{},
		size: 0,
	}
}

func (s *Stack) Pop() *TreeNode {
	if s.size <= 0 {
		return nil
	}

	tNode := s.arr[s.size-1]
	s.arr = s.arr[:s.size-1]
	s.size--

	return tNode
}

func (s *Stack) Push(node *TreeNode) {
	s.arr = append(s.arr, node)
	s.size++
}

func (s *Stack) IsEmpty() bool {
	if s.size == 0 {
		return true
	}

	return false
}

func preorderTraversal(root *TreeNode) []int {
	res := []int{}
	preorder(root, &res)

	return res
}

func preorder(root *TreeNode, res *[]int) {
	if root == nil {
		return
	}

	*res = append(*res, root.Val)

	preorder(root.Left, res)

	preorder(root.Right, res)
}

func preorderTraversal2(root *TreeNode) []int {
	res := []int{}
	stack := NewStack()

	if root == nil {
		return res
	}

	stack.Push(root)

	for !stack.IsEmpty() {
		node := stack.Pop()

		res = append(res, node.Val)

		if node.Right != nil {
			stack.Push(node.Right)
		}

		if node.Left != nil {
			stack.Push(node.Left)
		}
	}

	return res
}
